Add conventional commits plugin blog (#4475)
jenkins-infra/jenkins.io: content/blog/2021/07/2021-07-29-introducing-conventional-commits-plugin-for-jenkins.adoc

---
:layout: post
:title: "Introducing the Conventional Commits Plugin for Jenins"
:tags:
- gsoc
- gsoc2021
- conventionalcommits
- plugin
:author: adi10hero
:sig: gsoc
opengraph:
  image: /images/gsoc/2021/conventional-commits/conventionalCommitsPluginForJenkins.png
links:
  discourse: https://community.jenkins.io/t/introducing-the-conventional-commits-plugin/272
---

image:/images/gsoc/2021/conventional-commits/conventionalCommitsPluginForJenkins.png[GSoC, height=420, role=center, float=center]

> The conventional commits plugin is Google Summer of Code project. Special thanks to the mentors Gareth Evans, Kristin Whetstone, Olivier Vernin and Alan Buradezwic 

Reviewer: Asciidoc is easier to code read when we use sentence per line.  That's different from markdown where sentence per line would mean every line has a line break.
Also corrected spelling and added a word
```suggestion
The conventional commits plugin is a Google Summer of Code project.
Special thanks to the mentors Gareth Evans, Kristin Whetstone, Olivier Vernin and Allan Burdajewicz.
```

=== What are Conventional Commits

According to the link:https://www.conventionalcommits.org/[official website], conventonal commits are, "A specification for adding human and machine readable meaning to commit messages."

Conventional commits are a lightweight convention on top of commit messages.

The following table shows major structural elements offered by the conventional commits convention.

|===
|Structural Element|Example

|Chore
|`chore: improve logging`

|Fix
|`fix: minor bug fix`

|Feat
|`feat: add a new feature`

|Breaking Change
|`BREAKING CHANGE: reimplement`
|===

=== Why Conventional Commits

As the CI/CD world is moving more towards complete automation and minimal human interaction, the ability to fully automate a release is desired.
Conventional Commits enable the use of automated systems on top of commit messages.
These systems can "truly" automate a release with almost no human interaction. 

> The convention dovetails with link:https://semver.org/[SemVer]. Let's take an example, a maven project is currently versioned at `1.2.0`. The following table shows how conventional commits would bump the verison depending on the type of the commit.

Reviewer: Minor spelling changes for your consideration.
```suggestion
The convention dovetails with link:https://semver.org/[semantic versioning].
Let's take an example, a maven project is currently versioned at `1.2.0`.
The following table shows how conventional commits would bump the version depending on the type of the commit.
```

|===
|Commit Message|Version Bump|SemVer Equivalent

|`chore: improve logging`
|`1.2.0` -> `1.2.0`
|No version bump

|`fix: minor bug fix`
|`1.2.0` -> `1.2.1`
|Increament in the patch version

|`feat: add a new feature`
|`1.2.0` -> `1.3.0`
|Increment in the minor version

|`BREAKING CHANGE: reimplement`
|`1.2.0` -> `2.0.0`
|Increment in the major version
|===

=== The Conventional Commits Plugin

The conventional commits plugin is a Jenkins plugin to programatically determine the next semantic version of a git repository using:

- Last tagged version 
- Commit message log

==== How it works?

> The plugin will read the commit messages from the lastest tag till the lastest commit and determine what would be the next (Semantic) Version for that particular project.

Reviewer: I believe that adoc prefers an empty line after a heading.  Optional, especially because I may be wrong.
```suggestion

The plugin will read the commit messages from the lastest tag till the lastest commit and determine what would be the next (Semantic) Version for that particular project.
```

==== Supported Project Types?

> Currently the plugin supports the following project types:

Reviewer: I'm not sure what it means to "support" a project type.  Does the conventional commit plugin invoke the build tools (maven, gradle, make) as part of its execution?  Optionally would be nice to have more explanation of the meaning of "supporting" a project type.
```suggestion

Currently the plugin supports the following project types:
```

Author: Got it, I will try to expand more on this.

To answer your question, yes it (mostly) does invoke the build tools as a part of the execution. It depends on the type of project.

- Maven
- Gradle
- Make
- Python
- Helm
- Node (NPM)

==== How to request a project type support? 
Please feel free to open an issue on the link:https://github.com/jenkinsci/conventional-commits-plugin/[GitHub repository] of the plugin.

=== How to use the plugin

> Recommended way of using the plugin is to add a step in a Jenkins Pipeline Project.

Reviewer: Could you insert the example step here so that the reader can see it?
```suggestion
Recommended way of using the plugin is to add a step in a Jenkins Pipeline Project.
```

Author: Oh yes sure. Thank you for the suggestion 😄

The plugin is released on every feature using JEP-229.

> The plugin is available to download from the link:https://plugins.jenkins.io/conventional-commits[plguins site].

Reviewer: Spelling fix
```suggestion
The plugin is available to download from the link:https://plugins.jenkins.io/conventional-commits[plugins site].
```

=== Demo

You can watch the plugin in action in a demo presented at the GSoC Midterm Presentations

video::https://youtu.be/_D0hiA1Cgz8?t=3218[youtube,width=800,height=420]

=== Next Steps

- Support for pre-release information. Example: `1.0.0-alpha`, `1.0.0-beta`, etc
- Support for build metadata. Example: `1.0.0-beta+exp.sha.5114f85`
- Optionally writing the calcuated "Next Version" into the project's configuration file. Example: `pom.xml` for a maven project, `setup.py` for python.

=== Feedback

We would love to hear your feedback & suggestions for the plugin.

> Please reach out on the plugin's link:https://github.com/jenkinsci/conventional-commits-plugin[GitHub] repository, the link:https://gitter.im/jenkinsci/conventional-commits-plugin[Gitter] channel or start a dicussions on discourse.

Reviewer: ```suggestion
Please reach out on the plugin's link:https://github.com/jenkinsci/conventional-commits-plugin[GitHub] repository, the link:https://gitter.im/jenkinsci/conventional-commits-plugin[Gitter] channel or start a discussion on link:https://community.jenkins.io[community.jenkins.io].
```

As the CI/CD world is moving more towards complete automation and minimal human interaction, the ability to fully automate a release is desired.
Conventional Commits enable the use of automated systems on top of commit messages.
These systems can "truly" automate a release with almost no human interaction. 

The convention dovetails with link:https://semver.org/[semantic versioning].
Let's take an example, a maven project is currently versioned at `1.2.0`.
The following table shows how conventional commits would bump the version depending on the type of the commit.

|===
|Commit Message|Version Bump|SemVer Equivalent

|`chore: improve logging`
|`1.2.0` -> `1.2.0`
|No version bump

|`fix: minor bug fix`
|`1.2.0` -> `1.2.1`
|Increament in the patch version

|`feat: add a new feature`
|`1.2.0` -> `1.3.0`
|Increment in the minor version

|`BREAKING CHANGE: reimplement`
|`1.2.0` -> `2.0.0`
|Increment in the major version
|===

=== The Conventional Commits Plugin

The conventional commits plugin is a Jenkins plugin to programatically determine the next semantic version of a git repository using:

- Last tagged version 
- Commit message log

==== How it works?

> The plugin will read the commit messages from the lastest tag till the lastest commit and determine what would be the next (Semantic) Version for that particular project.

Reviewer: Spelling fixes
```suggestion
The plugin will read the commit messages from the latest tag till the latest commit and determine what would be the next semantic version for that particular project.
```

Author: Made the changes.
I am sorry for being clumsy with the spelling, I should have checked :(

Conventional commits are a lightweight convention on top of commit messages.

The following table shows major structural elements offered by the conventional commits convention.

|===
|Structural Element|Example

|Chore
|`chore: improve logging`

|Fix
|`fix: minor bug fix`

|Feat
|`feat: add a new feature`

|Breaking Change
|`BREAKING CHANGE: reimplement`
|===

=== Why Conventional Commits

As the CI/CD world is moving more towards complete automation and minimal human interaction, the ability to fully automate a release is desired.
Conventional Commits enable the use of automated systems on top of commit messages.
These systems can "truly" automate a release with almost no human interaction. 

The convention dovetails with link:https://semver.org/[semantic versioning].
Let's take an example, a maven project is currently versioned at `1.2.0`.
The following table shows how conventional commits would bump the version depending on the type of the commit.

|===
|Commit Message|Version Bump|SemVer Equivalent

|`chore: improve logging`

> |`1.2.0` -> `1.2.0`

Reviewer 2: I would have expected this to bump to `1.2.1` - although, with JEP-229 a release would not be created automatically.

|Increment in the minor version

|`BREAKING CHANGE: reimplement`
|`1.2.0` -> `2.0.0`
|Increment in the major version
|===

=== The Conventional Commits Plugin

The conventional commits plugin is a Jenkins plugin to programatically determine the next semantic version of a git repository using:

- Last tagged version 
- Commit message log
- Current version of the project

==== How it works?

The plugin will read the commit messages from the latest tag or the current version of the project till the latest commit.
Using this information it will determine what would be the next semantic Version for that particular project.

==== Supported Project Types?

Currently the plugin can read the current version from various configuration files of the following project types:

|===
|Project Type|Configuration File(s) Read

|Maven
|pom.xml

|Gradle
|build.gradle

|Make
|Makefile

|Python
|
setup.py

setup.cfg

pyproject.toml

|Helm
|Charts.yml

|Node (NPM)
|package.json
|===

==== How to request a project type support? 

Please feel free to open an issue on the link:https://github.com/jenkinsci/conventional-commits-plugin/[GitHub repository] of the plugin.

=== How to use the plugin

> Recommended way of using the plugin is to add a step in a Jenkins Pipeline project.

Reviewer 3: I think what I'd do is note that `nextVersion()` is the specific pipeline step.  I'd also mention that you can see the pipeline help in the Snippet Generator.
